atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetEmptyLatestValues.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.sweep;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+
+public class GetEmptyLatestValues {
+
+    private final List<CellWithTimestamps> cellTimestamps;
+    private final ValuesLoader valuesLoader;
+    private final TableReference table;
+    private final long maxTimestampExclusive;
+    private final int batchSize;
+
+    public GetEmptyLatestValues(
+            List<CellWithTimestamps> cellTimestamps,
+            ValuesLoader valuesLoader,
+            TableReference table,
+            long maxTimestampExclusive,
+            int batchSize) {
+        this.cellTimestamps = cellTimestamps;
+        this.valuesLoader = valuesLoader;
+        this.table = table;
+        this.maxTimestampExclusive = maxTimestampExclusive;
+        this.batchSize = batchSize;
+    }
+
+    /**
+     * Returns the subset of {@link Cell}s whose latest values prior to {@code maxTimestampExclusive} are empty.
+     */
+    public Set<Cell> execute() {
+        Set<Cell> result = Sets.newHashSet();
+        for (List<CellWithTimestamps> batch : Iterables.partition(cellTimestamps, batchSize)) {
+            result.addAll(getSingleBatch(batch));
+        }
+        return result;
+    }
+
+    private Set<Cell> getSingleBatch(List<CellWithTimestamps> batch) {
+        Set<Cell> cells = batch.stream()
+                .map(CellWithTimestamps::cell)
+                .collect(Collectors.toSet());
+
+        Map<Cell, Value> valuesByCell = valuesLoader.getValues(table, cells, maxTimestampExclusive);
```

Comment:
...So we still load all values if the strategy is THOROUGH? What is exactly the point of this change then?

Also, have you seen the implementation of loadWithTs? (Which is the code behind the ValuesLoader) It makes a separate request for each unique column name. For wide rows, this code will perform very poorly.

Reply:
The difference is that we only load values for the latest cell (as opposed to every cell).

ValuesLoader was meant to be a conservative impl pending benchmarks. The main goal was to not OOM or overfetch. But agree that we should revisit this.

Reply:
But in the case of an (almost) clean table, every cell will basically be the latest cell. Which means that in the worst case, the OOM concern is still valid. (And this worst case is pretty common - most tables should be almost clean if sweep is doing its job well).

Also, internally Cassandra will still fetch all values from disk, so we are not saving much by trying to load the latest values separately.

(By the way, you can run benchmarks manually before you merge, or even push a change - it's pretty easy, ping me on Slack if you need help)